dockstore/dockstore
# Support Userless workflows

**Description**
1. Allows workflows to be created from a .dockstore.yml if there is no corresponding user on Dockstore.
2. Also look for more users to make owners of the workflow upon creation.

***Why***
Some workflows are new via a .dockstore.yml, but the push is done by a bot not registered in Dockstore, so the workflow doesn't appear. 

Why did we not support userless workflows? This was explicitly disallowed, with a test, when services were added in #2574, and the code was carried over to workflows. I could find no other reasons for it, although I can see how it makes sense. There are no implications as far as I can tell by allowing workflows without a user.

***Users in a GitHub webhook push event***
Following is extracted from [here](https://docs.github.com/en/webhooks/webhook-events-and-payloads#push):
There are potentially many unique GitHub users in a GitHub event:
1. The sender - the only user we've been looking at so far. It's the user that triggered the event. In our team's typical usage it's the same as the person who committed and pushed, but I think it could be, for example, a bot that created and pushed a branch.
2. For a commit, there is an author and a committer.
    1. The author is the user who wrote the code.
    2. The committer is the user who commited the code, who is not necessarily the same as the author.
        1. Example 1, I edit a file in the GitHub web UI. I will be the author, the user `web-flow` will be the committer.
        2. Example 2, I commit file to a branch. I am the author and the committer. Denis cherry-picks my commit to another branch. I am the author of the cherry-picked commit, but Denis is the committer.

The payload has a `head_commit`, which could be null, and a `commits` array, which could be empty. I'm not clear how far back in time the commits array can go, and when `head_commit` is set.

***Looking for additional users***
Note: we could drop this part if we think it's risky/slow.

Even though the idea is to support userless workflows, it would be nice it we could auto-discover some users if the sender is not in Dockstore. We look through the committers and authors to see if we can add any of them as owners.

I'm concerned that an author/committer may not have current permissions (for example a PR from a fork might get merged, but it doesn't mean the author has permissions), so we do a check to be safe. This check can be slow.

Why don't we just query the repo for its owners? Because the GitHub app doesn't have permissions.

**Review Instructions**
This will be tricky, I think you can simulate it by unlinking your GitHub account:
1. Ensure you have a repo configured with the GitHub App
2. Unlink GitHub from your Dockstore account
3. Add a new workflow, with publish set to true, in the .dockstore.yml
4. Push it
5. After a couple of minutes the workflow should appear in search
6. Relink your GitHub account in Dockstore.
7. You should not see the new workflow in My Workflows
8. Click Discover Workflows.
9. You should now see the new workflow in Dockstore.

**Issue**
SEAB-5994

**Security and Privacy**

If there are any concerns that require extra attention from the security team, highlight them here and check the box when complete. 


e.g. Does this change...
* Any user data we collect, or data location?
* Access control, authentication or authorization?
* Encryption features?

Please make sure that you've checked the following before submitting your pull request. Thanks!
